File: other-programming-languages/haskell/2024/2024-004/main.hs
import System.IO
import Data.List (transpose)

-- Define the target word and its length
targetWord :: String
targetWord = "XMAS"

wordLength :: Int
wordLength = length targetWord

-- Directions for moving in the grid
directions :: [(Int, Int)]
directions = 
    [ (0, 1)   -- Right
    , (1, 0)   -- Down
    , (1, 1)   -- Diagonal-right-down
    , (1, -1)  -- Diagonal-left-down
    , (0, -1)  -- Left
    , (-1, 0)  -- Up
    , (-1, -1) -- Diagonal-left-up
    , (-1, 1)  -- Diagonal-right-up
    ]

-- Function to check if a word exists in a given direction
checkWord :: [[Char]] -> (Int, Int) -> (Int, Int) -> Bool
checkWord grid (x, y) (dx, dy) =
    let positions = take wordLength [(x + i * dx, y + i * dy) | i <- [0..]]
        isValid (r, c) = r >= 0 && c >= 0 && r < length grid && c < length (head grid)
        chars = [grid !! r !! c | (r, c) <- positions, isValid (r, c)]
    in length chars == wordLength && chars == targetWord

-- Count all occurrences of the target word
countOccurrences :: [[Char]] -> Int
countOccurrences grid =
    length [()
        | r <- [0..length grid - 1]
        , c <- [0..length (head grid) - 1]
        , dir <- directions
        , checkWord grid (r, c) dir]

-- Function to count all X-MAS patterns
countXmasPatterns :: [[Char]] -> Int
countXmasPatterns grid =
    length [()
        | r <- [1..length grid - 2]
        , c <- [1..length (head grid) - 2]
        , let center = grid !! r !! c
              topLeft = grid !! (r - 1) !! (c - 1)
              topRight = grid !! (r - 1) !! (c + 1)
              bottomLeft = grid !! (r + 1) !! (c - 1)
              bottomRight = grid !! (r + 1) !! (c + 1)
        , center == 'A'
        , (topLeft, topRight, bottomLeft, bottomRight) `elem`
            [ ('M', 'S', 'M', 'S')
            , ('S', 'M', 'S', 'M')
            , ('M', 'M', 'S', 'S')
            , ('S', 'S', 'M', 'M')
            ]]

-- Main function to process the input file and compute results
main :: IO ()
main = do
    -- Read the grid from the input file
    contents <- readFile "input.txt"
    let grid = lines contents

    -- Count occurrences of the word "XMAS"
    let wordCount = countOccurrences grid
    putStrLn $ "count is " ++ show wordCount

    -- Count all X-MAS patterns
    let xmasPatternCount = countXmasPatterns grid
    putStrLn $ "total xmas patterns is " ++ show xmasPatternCount

File: other-programming-languages/Swift/2024/2024-004/main.swift
import Foundation

// Function to read the file and return the grid as an array of strings
func readFile(_ filePath: String) -> [String] {
    do {
        let contents = try String(contentsOfFile: filePath, encoding: .utf8)
        return contents.split(separator: "\n").map { String($0) }
    } catch {
        print("Error reading file: \(error)")
        return []
    }
}

// Function to check if a word exists in a given direction
func checkWord(grid: [[Character]], word: String, x: Int, y: Int, dx: Int, dy: Int) -> Bool {
    let rows = grid.count
    let cols = grid[0].count
    for (i, char) in word.enumerated() {
        let nx = x + i * dx
        let ny = y + i * dy
        if nx < 0 || ny < 0 || nx >= rows || ny >= cols || grid[nx][ny] != char {
            return false
        }
    }
    return true
}

// Count all occurrences of the target word
func countOccurrences(grid: [[Character]], word: String) -> Int {
    let directions = [
        (0, 1),   // Right
        (1, 0),   // Down
        (1, 1),   // Diagonal-right-down
        (1, -1),  // Diagonal-left-down
        (0, -1),  // Left
        (-1, 0),  // Up
        (-1, -1), // Diagonal-left-up
        (-1, 1)   // Diagonal-right-up
    ]
    let rows = grid.count
    let cols = grid[0].count
    var count = 0

    for r in 0..<rows {
        for c in 0..<cols {
            for (dx, dy) in directions {
                if checkWord(grid: grid, word: word, x: r, y: c, dx: dx, dy: dy) {
                    count += 1
                }
            }
        }
    }

    return count
}

// Count all X-MAS patterns
func countXmasPatterns(grid: [[Character]]) -> Int {
    let rows = grid.count
    let cols = grid[0].count
    var count = 0

    for r in 1..<rows - 1 {
        for c in 1..<cols - 1 {
            let center = grid[r][c]
            let topLeft = grid[r - 1][c - 1]
            let topRight = grid[r - 1][c + 1]
            let bottomLeft = grid[r + 1][c - 1]
            let bottomRight = grid[r + 1][c + 1]

            if center == "A" {
                if (topLeft == "M" && topRight == "S" && bottomLeft == "M" && bottomRight == "S") ||
                   (topLeft == "S" && topRight == "M" && bottomLeft == "S" && bottomRight == "M") ||
                   (topLeft == "M" && topRight == "M" && bottomLeft == "S" && bottomRight == "S") ||
                   (topLeft == "S" && topRight == "S" && bottomLeft == "M" && bottomRight == "M") {
                    count += 1
                }
            }
        }
    }

    return count
}

// Main logic
let filePath = "input.txt"
let gridStrings = readFile(filePath)
let grid = gridStrings.map { Array($0) }
let targetWord = "XMAS"

// Count occurrences of the word "XMAS"
let wordCount = countOccurrences(grid: grid, word: targetWord)
print("count is \(wordCount)")

// Count all X-MAS patterns
let xmasPatternCount = countXmasPatterns(grid: grid)
print("total xmas patterns is \(xmasPatternCount)")
